finquant/data_types.py

```diff
@@ -0,0 +1,74 @@
+"""
+``finquant.data_types`` Module
+
+This module defines type aliases and utility functions for working with arrays, data frames,
+and various numeric types in Python, utilizing the 'numpy', 'numpy.typing', and 'pandas' libraries.
+
+Generic List Element Type
+-------------------------
+- ``ELEMENT_TYPE``: A type alias representing a generic element type
+
+Array/List-Like Types
+---------------------
+- ``ARRAY_OR_LIST``: A type alias representing either a NumPy ``ndarray`` or a Python ``List``.
+- ``ARRAY_OR_DATAFRAME``: A type alias representing either a NumPy ``ndarray`` or a pandas ``DataFrame``.
+- ``ARRAY_OR_SERIES``: A type alias representing either a NumPy ``ndarray`` or a pandas ``Series``.
+- ``SERIES_OR_DATAFRAME``: A type alias representing either a pandas ``Series`` or a pandas ``DataFrame``.
+
+Numeric Types
+-------------
+- ``FLOAT``: A type alias representing either a NumPy floating-point number or a Python float.
+- ``INT``: A type alias representing either a NumPy integer or a Python int.
+- ``NUMERIC``: A type alias representing either an ``INT`` or a ``FLOAT``.
+
+String/Datetime Types
+---------------------
+- ``STRING_OR_DATETIME``: A type alias representing either a Python string or a ``datetime.datetime`` object.
+
+Dependencies
+------------
+This module requires the following external libraries:
+
+- ``numpy`` (imported as ``np``)
+- ``pandas`` (imported as ``pd``)
+
+Usage Example
+-------------
+
+.. code-block:: python
+
+    from finquant.data_types import ARRAY_OR_DATAFRAME, NUMERIC
+    # Use the defined type aliases
+    def process_data(data: ARRAY_OR_DATAFRAME) -> FLOAT:
+        # Process the data and return a floating point number
+        return 5.0
+
+"""
+# pylint: disable=C0103
+
+
+from datetime import datetime
+from typing import Any, KeysView, List, TypeVar, Union
+
+import numpy as np
+import pandas as pd
+
+# Generic List Element Type
+ELEMENT_TYPE = TypeVar("ELEMENT_TYPE")
+
+# Type Aliases:
+ARRAY_OR_LIST = Union[np.ndarray[ELEMENT_TYPE, Any], List[ELEMENT_TYPE]]
```

Comment:
Have a quick look at the custom defined data types here (up to `STRING_OR_DATETIME`.